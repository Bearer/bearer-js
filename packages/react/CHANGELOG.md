# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.95.1](https://github.com/Bearer/bearer/compare/v0.95.0...v0.95.1) (2019-02-20)

**Note:** Version bump only for package @bearer/react





# [0.95.0](https://github.com/Bearer/bearer/compare/v0.94.1...v0.95.0) (2019-02-15)

**Note:** Version bump only for package @bearer/react





## [0.94.1](https://github.com/Bearer/bearer/compare/v0.94.0...v0.94.1) (2019-02-12)

**Note:** Version bump only for package @bearer/react





# [0.94.0](https://github.com/Bearer/bearer/compare/v0.93.4...v0.94.0) (2019-02-12)

**Note:** Version bump only for package @bearer/react





# [0.93.0](https://github.com/Bearer/bearer/compare/v0.92.1...v0.93.0) (2019-02-05)


### Features

* bump all versions ([4fb7615](https://github.com/Bearer/bearer/commit/4fb7615))





## [0.92.1](https://github.com/Bearer/bearer/compare/v0.92.0...v0.92.1) (2019-02-01)

**Note:** Version bump only for package @bearer/react





# [0.92.0](https://github.com/Bearer/bearer/compare/v0.91.8...v0.92.0) (2019-01-31)

**Note:** Version bump only for package @bearer/react





## [0.91.5](https://github.com/Bearer/bearer/compare/v0.91.4...v0.91.5) (2019-01-29)

**Note:** Version bump only for package @bearer/react





# [0.91.0](https://github.com/Bearer/bearer/compare/v0.90.6...v0.91.0) (2019-01-24)

**Note:** Version bump only for package @bearer/react





## [0.90.2](https://github.com/Bearer/bearer/compare/v0.90.1...v0.90.2) (2019-01-22)

**Note:** Version bump only for package @bearer/react





## [0.90.1](https://github.com/Bearer/bearer/compare/v0.90.0...v0.90.1) (2019-01-22)

**Note:** Version bump only for package @bearer/react





<a name="0.90.0"></a>
# [0.90.0](https://github.com/Bearer/bearer/compare/v0.89.2...v0.90.0) (2019-01-22)

**Note:** Version bump only for package @bearer/react





<a name="0.89.0"></a>
# [0.89.0](https://github.com/Bearer/bearer/compare/v0.88.0...v0.89.0) (2019-01-17)

**Note:** Version bump only for package @bearer/react





<a name="0.88.0"></a>
# [0.88.0](https://github.com/Bearer/bearer/compare/v0.87.3...v0.88.0) (2019-01-17)

**Note:** Version bump only for package @bearer/react





<a name="0.87.0"></a>
# [0.87.0](https://github.com/Bearer/bearer/compare/v0.86.2...v0.87.0) (2019-01-15)

**Note:** Version bump only for package @bearer/react





<a name="0.86.0"></a>
# [0.86.0](https://github.com/Bearer/bearer/compare/v0.85.4...v0.86.0) (2019-01-03)

**Note:** Version bump only for package @bearer/react





<a name="0.85.4"></a>
## [0.85.4](https://github.com/Bearer/bearer/compare/v0.85.3...v0.85.4) (2019-01-02)


### Bug Fixes

* **react:** allow different int host for react applications ([9024547](https://github.com/Bearer/bearer/commit/9024547))
* **react:** make it discoverable on npm ([30923f5](https://github.com/Bearer/bearer/commit/30923f5))





<a name="0.85.0"></a>
# [0.85.0](https://github.com/Bearer/bearer/compare/v0.84.2...v0.85.0) (2018-12-27)

**Note:** Version bump only for package @bearer/react





<a name="0.84.2"></a>
## [0.84.2](https://github.com/Bearer/bearer/compare/v0.84.1...v0.84.2) (2018-12-27)

**Note:** Version bump only for package @bearer/react





<a name="0.84.1"></a>
## [0.84.1](https://github.com/Bearer/bearer/compare/v0.84.0...v0.84.1) (2018-12-27)

**Note:** Version bump only for package @bearer/react





<a name="0.84.0"></a>
# [0.84.0](https://github.com/Bearer/bearer/compare/v0.83.5...v0.84.0) (2018-12-21)

**Note:** Version bump only for package @bearer/react





<a name="0.83.4"></a>
## [0.83.4](https://github.com/Bearer/bearer/compare/v0.83.3...v0.83.4) (2018-12-21)

**Note:** Version bump only for package @bearer/react





<a name="0.83.3"></a>
## [0.83.3](https://github.com/Bearer/bearer/compare/v0.83.2...v0.83.3) (2018-12-21)

**Note:** Version bump only for package @bearer/react





<a name="0.83.0"></a>
# [0.83.0](https://github.com/Bearer/bearer/compare/v0.82.0...v0.83.0) (2018-12-20)

**Note:** Version bump only for package @bearer/react





<a name="0.82.0"></a>
# [0.82.0](https://github.com/Bearer/bearer/compare/v0.81.9...v0.82.0) (2018-12-19)


### Features

* **security:** introduce security package ([#441](https://github.com/Bearer/bearer/issues/441)) ([aae485f](https://github.com/Bearer/bearer/commit/aae485f))





<a name="0.81.7"></a>
## [0.81.7](https://github.com/Bearer/bearer/compare/v0.81.6...v0.81.7) (2018-12-19)


### Bug Fixes

* **react:** make the package public ([d45c6a6](https://github.com/Bearer/bearer/commit/d45c6a6))





<a name="0.81.6"></a>
## [0.81.6](https://github.com/Bearer/bearer/compare/v0.81.5...v0.81.6) (2018-12-19)


### Bug Fixes

* **react:** fix issue where context would not update + expose context ([#437](https://github.com/Bearer/bearer/issues/437)) ([811c4b4](https://github.com/Bearer/bearer/commit/811c4b4))





<a name="0.81.5"></a>
## [0.81.5](https://github.com/Bearer/bearer/compare/v0.81.4...v0.81.5) (2018-12-18)


### Bug Fixes

* **react:** fix build scripts and package naming ([#436](https://github.com/Bearer/bearer/issues/436)) ([3b3ab50](https://github.com/Bearer/bearer/commit/3b3ab50))





<a name="0.81.3"></a>
## [0.81.3](https://github.com/Bearer/bearer-react/compare/v0.81.2...v0.81.3) (2018-12-18)

**Note:** Version bump only for package bearer-react
