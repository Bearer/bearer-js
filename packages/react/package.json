{
  "name": "@bearer/react",
  "private": false,
  "version": "0.92.1",
  "description": "React tooling for bearer.sh components",
  "keywords": [
    "react",
    "bearer"
  ],
  "author": "Bearer Team<engineering@bearer.sh>",
  "homepage": "https://github.com/Bearer/bearer#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__test__"
  },
  "files": [
    "lib/"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Bearer/bearer.git"
  },
  "bugs": {
    "url": "https://github.com/Bearer/bearer/issues"
  },
  "devDependencies": {
    "@bearer/package-init": "^0.78.0",
    "@commitlint/cli": "^7.3.2",
    "@commitlint/config-conventional": "^7.3.1",
    "@oclif/tslint": "^3.1.1",
    "@types/jest": "^23.3.13",
    "@types/lodash.kebabcase": "^4.1.4",
    "@types/react": "^16.7.17",
    "@types/react-helmet": "^5.0.7",
    "@types/react-test-renderer": "^16.0.3",
    "commitlint": "^7.3.2",
    "cz-conventional-changelog": "^2.1.0",
    "husky": "^1.3.1",
    "jest": "^23.6.0",
    "lint-staged": "^8.1.0",
    "prettier": "^1.16.1",
    "react": "^16.6.0-0",
    "react-test-renderer": "^16.6.3",
    "ts-jest": "^23.10.5",
    "tslint": "^5.12.1",
    "tslint-config-prettier": "^1.17.0",
    "typescript": "^3.2.4"
  },
  "peerDependencies": {
    "react": "^16.6.0-0"
  },
  "dependencies": {
    "lodash.kebabcase": "^4.1.1",
    "react-helmet": "^5.2.0"
  },
  "config": {
    "commitizen": {
      "path": "./node_modules/cz-conventional-changelog"
    }
  },
  "husky": {
    "hooks": {
      "pre-commit": "lint-staged",
      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS"
    }
  },
  "scripts": {
    "clean": "rm -rf lib/",
    "start": "yarn build -w",
    "build": "tsc -p tsconfig.build.json",
    "prepare": "yarn build",
    "prepack": "yarn clean && NODE_ENV=production yarn build",
    "test": "jest",
    "test:ci": "jest --coverage --runInBand"
  }
}
