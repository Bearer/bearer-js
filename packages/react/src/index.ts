import BearerProvider, {BearerContext} from './bearer-provider'
import fromBearer from './from-bearer'
import BearerLoader from './bearer-loader'

export { 
  BearerLoader,
  BearerProvider,
  BearerContext,
  fromBearer
}
