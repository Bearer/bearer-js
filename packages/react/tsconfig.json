{
  "compilerOptions": {
    "declaration": true,
    "target": "es5",
    "module": "commonjs",
    "strict": true,
    "esModuleInterop": true,
    "outDir": "lib/",
    "jsx": "react",
    "lib": ["es2015", "dom"]
  },
  "include": ["src/"]
}
