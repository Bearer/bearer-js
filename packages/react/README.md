# @bearer/react

React tooling for bearer.sh components

## Installation

Install dependencies

```bash
yarn add @bearer/js @bearer/react
```

## fromBearer

This component takes html tag name of a component as well as optional type information to produce a react component that can we re-used throughout an application.

```tsx
import * as React from 'react'
const MyText = fromBearer<{ text?: string }>('bearer-my-text')

class ReactComponent extends React.Component {
  render() {
    return <MyText text="hello world" />
  }
}
```

Output events can be handled by simply adding the eventname as a prop on the component and adding a basic handler

```tsx
<Share bearer-uuid-feature-shared={this.onShared} />
```

For these tags to work correctly they must have a parent `BearerProvider`

## BearerProvider

This component maintains a shared state for a group of components as well as adding the bearer script tags to the page. This tag can be added at any level of the application as is convenient for the implementation

## Example Use

we will use [this scenario](https:// app.bearer.sh/scenarios/6d29c4-share-slack-beta-4/preview) for our examples:

First on the preview page obtain a setup-id from the setup components so we do not need to include these in our page. Next lets define our components in a constants file:

`bearer.ts`

```TS
import { fromBearer } from '@bearer/react'

export const ChannelSelect = fromBearer('bearer-6d29c4-share-slack-beta-4-channel-action')
export const Share = fromBearer<{message?:string, text?:string}>('bearer-6d29c4-share-slack-beta-4-feature-action')
export const SlackConnect = fromBearer('bearer-6d29c4-share-slack-beta-4-connect-action')
```

We can include the `BearerProvider` at any level but in this example lets use it all together in the same component

`slack-share-component.tsx`

```tsx
import * as React from 'react'
import { BearerProvider } from '@bearer/react'
import { SlackConnect, ChannelSelect, Share } from '../../constants/bearer'

export default class SlackShareSetup extends React.Component {
  public render() {
    const intialContext = { 'setup-id': 'SETUP_ID_SAMPLE' }
    return (
      <BearerProvider clientId="YOUR_CLIENT_ID" initialContext={intialContext}>
        <SlackConnect />
        <ChannelSelect />
        <Share message="hello world!" text="Test!" bearer-6d29c4-share-slack-beta-4-feature-shared={this.onShared} />
      </BearerProvider>
    )
  }
}
```

This allows to share messages but what if we want to persist the users setup information. @bearer/react provides two methods to acomplish this.

### Using onUpdate

`BearerProvider` has an `onUpdate`callback we can hook into which is called every time data changes within the provider

`slack-share-component.tsx`

```tsx
import * as React from 'react'
import { BearerProvider } from '@bearer/react'
import { SlackConnect, ChannelSelect, Share } from '../../constants/bearer'

export default class SlackShareSetup extends React.Component {
  public render() {
    const intialContext = { 'setup-id': 'SETUP_ID_SAMPLE' }
    return (
      <BearerProvider
        clientId="YOUR_CLIENT_ID"
        initialContext={intialContext}
        onUpdate={(data: any) => {
          this.setState({ data })
        }}
      >
        <SlackConnect />
        <ChannelSelect />
        <Share message="hello world!" text="Test!" bearer-6d29c4-share-slack-beta-4-feature-shared={this.onShared} />
        <button onClick={this.handleSave}>Save Setup</button>
      </BearerProvider>
    )
  }
  private handleSave = () => console.log('handleSave', this.state.data)
}
```

### using a context consumer

Internally `BearerProvider` uses the [react context API](https:// reactjs.org/docs/context.html). For a more advaned but flexable method we can access the consumer directly via `BearerContext` and then use the currently set details as we wish.

`slack-share-component.tsx`

```tsx
import * as React from 'react'
import { BearerProvider, BearerContext } from '@bearer/react'
import { SlackConnect, ChannelSelect, Share } from '../../constants/bearer'

export default class SlackShareSetup extends React.Component {
  public render() {
    const intialContext = { 'setup-id': 'SETUP_ID_SAMPLE' }
    return (
      <BearerProvider clientId="YOUR_CLIENT_ID" initialContext={intialContext}>
        <SlackConnect />
        <ChannelSelect />
        <Share message="hello world!" text="Test!" bearer-6d29c4-share-slack-beta-4-feature-shared={this.onShared} />
        <BearerContext.Consumer>
          {context => <button onClick={this.handleSave(context.state)}>Save Setup</button>}
        </BearerContext.Consumer>
      </BearerProvider>
    )
  }
  private handleSave = (data: any) => () => {
    console.log('handleSave', data)
  }
}
```
