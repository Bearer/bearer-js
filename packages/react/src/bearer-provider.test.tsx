import * as React from 'react'
import * as Renderer from 'react-test-renderer'
import * as ShallowRenderer from 'react-test-renderer/shallow'

import DummyContextConsumer from '../__test__/utils/dummy-context-consumer'

import BearerProvider from './bearer-provider'

describe('BearerProvider', () => {
  const clientId = 'abc123'
  const initialContext = {
    'setup-id': '123'
  }

  it('renders basic provider tag', () => {
    const renderer = ShallowRenderer.createRenderer()
    const tree = renderer.render(<BearerProvider clientId={clientId} />)
    expect(tree).toMatchSnapshot()
  })

  it('allow custom integration host', () => {
    const renderer = ShallowRenderer.createRenderer()
    const tree = renderer.render(<BearerProvider clientId={clientId} intHost="https://int.bearer.sh/v2/" />)
    expect(tree).toMatchSnapshot()
  })

  it('default state is sent to context components', () => {
    const testRender = Renderer.create(
      <BearerProvider clientId={clientId} initialContext={initialContext}>
        <DummyContextConsumer />
      </BearerProvider>
    )
    expect(testRender.toJSON()).toMatchSnapshot()
  })

  it('updates are handled', () => {
    const changeHandler = jest.fn(_x => null)
    const testRender = Renderer.create(
      <BearerProvider clientId={clientId} initialContext={initialContext} onUpdate={changeHandler}>
        <DummyContextConsumer />
      </BearerProvider>
    )
    const stopPropagation = jest.fn(_x => null)
    const mockEvent = {
      stopPropagation,
      detail: { someRefrenceId: 'abc' }
    }
    testRender.root.findByType('button').props.onClick(mockEvent)
    expect(stopPropagation.mock.calls.length).toBe(1)
    expect(changeHandler.mock.calls.length).toBe(1)
    expect(changeHandler.mock.calls[0][0]).toEqual({
      ...initialContext,
      'some-refrence-id': 'abc'
    })
  })
})
