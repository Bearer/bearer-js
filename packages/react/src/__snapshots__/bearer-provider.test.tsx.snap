// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BearerProvider allow custom integration host 1`] = `
<Context.Provider
  value={
    Object {
      "bearer": e {
        "clientId": "abc123",
        "config": Object {
          "domObserver": true,
          "integrationHost": "https://integrations.bearer.sh/",
          "refreshDebounceDelay": 200,
          "secured": undefined,
        },
        "connect": [Function],
        "debounceRefresh": [Function],
        "disconnectObserver": [Function],
        "initialIntegrationLoading": [Function],
        "integration": [Function],
        "invoke": [Function],
        "loadMissingIntegrations": [Function],
        "observerCallback": [Function],
        "registerDomObserver": [Function],
        "registeredIntegration": [Function],
        "registeredIntegrations": Object {},
        "secured": undefined,
        "sendTags": [Function],
      },
      "handlePropUpdates": [Function],
      "state": Object {},
    }
  }
/>
`;

exports[`BearerProvider default state is sent to context components 1`] = `
Array [
  "{
  \\"setup-id\\": \\"123\\"
}",
  <button
    onClick={[Function]}
  />,
]
`;

exports[`BearerProvider renders basic provider tag 1`] = `
<Context.Provider
  value={
    Object {
      "bearer": e {
        "clientId": "abc123",
        "config": Object {
          "domObserver": true,
          "integrationHost": "https://int.bearer.sh",
          "refreshDebounceDelay": 200,
          "secured": undefined,
        },
        "connect": [Function],
        "debounceRefresh": [Function],
        "disconnectObserver": [Function],
        "initialIntegrationLoading": [Function],
        "integration": [Function],
        "invoke": [Function],
        "loadMissingIntegrations": [Function],
        "observerCallback": [Function],
        "registerDomObserver": [Function],
        "registeredIntegration": [Function],
        "registeredIntegrations": Object {},
        "secured": undefined,
        "sendTags": [Function],
      },
      "handlePropUpdates": [Function],
      "state": Object {},
    }
  }
/>
`;
